### ISSUE TITLE
False non-copyable report

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn-analyzers#7389]
### Analyzer

**Diagnostic ID**: RS0042: Unsupported use of non-copyable type 'Microsoft.AspNetCore.Razor.PooledObjects.PooledArrayBuilder<Microsoft.AspNetCore.Razor.Language.Syntax.InternalSyntax.SyntaxToken>' in 'LocalReference' operation

### Analyzer source

**NuGet Package**: Roslyn.Diagnostics.Analyzers

**Version**: 3.11.0-beta1.24170.2

### Describe the bug

When using a non-copyable struct in a list pattern, RS0042 erroneously fires.

### Steps To Reproduce

1. Create a collection type that supports list patterns, and mark it non-copyable.
2. Use that collection in a list pattern.

### Expected behavior

No errors

### Actual behavior

RS0042: Unsupported use of non-copyable type 'Microsoft.AspNetCore.Razor.PooledObjects.PooledArrayBuilder<Microsoft.AspNetCore.Razor.Language.Syntax.InternalSyntax.SyntaxToken>' in 'LocalReference' operation

**Comment by 333fred** [comment]
![image](https://github.com/user-attachments/assets/f0894c66-bf01-4098-9fb8-4ce9ed7e362f)
As an example.

**Comment by sharwell** [comment]
In reviewing this code on SharpLab, it looks like there are some cases where a shadow copy of the pattern variable is made. The first step of resolving this will be identifying cases where the value is _definitely known to not be copied_.

```csharp
using System;
using System.Collections;
using System.Collections.Generic;

public class C {
    C2 _value;

    public void M() {
        using var value = default(C2);
        if (value is [.., string s])
        {
            return;
        }
    }

    public void M2() {
        if (_value is [.., string s2])
        {
            return;
        }
    }
}

public struct C2 : IEnumerable<object>, IDisposable
{
    public int Count => 1;

    public object this[int index] => throw new NotImplementedException();

    public void Dispose()
    {
        throw new NotImplementedException();
    }

    public IEnumerator<object> GetEnumerator()
    {
        throw new NotImplementedException();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        throw new NotImplementedException();
    }
}
```

**Comment by 333fred** [comment]
I would allow it in all cases. A list paeis simply an enumeration: if that affects the state of the struct, then it's already broken.
